packages/realm_dart/lib/src/logging.dart

```diff
@@ -0,0 +1,182 @@
+        return false;
+      }
+      current = current._parent!;
+    }
+    return true;
+  }
+
+  // cache the toString result
+  late final String _toString = (_parent == null ? _name : '$_parent.$_name');
+  @override
+  String toString() => _toString;
+
+  static final _map = {for (final category in RealmLogCategory.values) category.toString(): category};
+  factory RealmLogCategory.fromString(String message) => _map[message]!;
+}
+
+class _LeafLogCategory extends RealmLogCategory {
+  _LeafLogCategory(super.name, RealmLogCategory super.parent) : super._();
+}
+
+class _RealmLogCategory extends RealmLogCategory {
+  _RealmLogCategory() : super._('Realm', null);
+
+  late final app = _LeafLogCategory('App', this);
+  late final sdk = _LeafLogCategory('SDK', this);
+  late final storage = _StorageLogCategory(this);
+  late final sync = _SyncLogCategory(this);
+}
+
+class _SyncLogCategory extends RealmLogCategory {
+  _SyncLogCategory(RealmLogCategory parent) : super._('Sync', parent);
+
+  late final client = _ClientLogCategory(this);
+  late final server = _LeafLogCategory('Server', this);
+}
+
+class _ClientLogCategory extends RealmLogCategory {
+  _ClientLogCategory(RealmLogCategory parent) : super._('Client', parent);
+
+  late final changeset = _LeafLogCategory('Changeset', this);
+  late final network = _LeafLogCategory('Network', this);
+  late final reset = _LeafLogCategory('Reset', this);
+  late final session = _LeafLogCategory('Session', this);
+}
+
+class _StorageLogCategory extends RealmLogCategory {
+  _StorageLogCategory(RealmLogCategory parent) : super._('Storage', parent);
+
+  late final notification = _LeafLogCategory('Notification', this);
+  late final object = _LeafLogCategory('Object', this);
+  late final query = _LeafLogCategory('Query', this);
+  late final transaction = _LeafLogCategory('Transaction', this);
+}
+
+/// Specifies the criticality level above which messages will be logged
+/// by the default sync client logger.
+/// {@category Realm}
+enum RealmLogLevel {
+  /// Log everything. This will seriously harm the performance of the
+  /// sync client and should never be used in production scenarios.
+  all(Level.ALL),
+
+  /// A version of [debug] that allows for very high volume output.
+  /// This may seriously affect the performance of the sync client.
+  trace(Level.FINEST),
+
+  /// Reveal information that can aid debugging, no longer paying
+  /// attention to efficiency.
+  debug(Level.FINER),
+
+  /// Same as [info], but prioritize completeness over minimalism.
+  detail(Level.FINE),
+
+  /// Log operational sync client messages, but in a minimalist fashion to
+  /// avoid general overhead from logging and to keep volume down.
+  info(Level.INFO),
+
+  /// Log errors and warnings.
+  warn(Level.WARNING),
+
+  /// Log errors only.
+  error(Level.SEVERE),
+
+  /// Log only fatal errors.
+  fatal(Level.SHOUT),
+
+  /// Turn off logging.
+  off(Level.OFF),
+  ;
+
+  static const logToValues = [
```

Comment:
Should this be public?

Reply:
If not, it should. simply be removed, as it is only used by the tests.